Add 'Find Sequence' to `Span`s, and consolidate negative indexing behavior

ColinSORourke opened the pull request:
- Maintainer edit, requires https://github.com/godotengine/godot/pull/106342

This Pull Request contains 4 things, using Ivorforce's Spans to improve String & TypedArray Find functions

1. Added Span::Find_Seq function to the Span class in Span.h
2. Refactored String::find to call Span::Find_Seq
3. Added Find_Seq function to Vector.h, and exposed to all PackedArrays in Variant_Call.cpp
4. Added Documentation for Find_Sequence for the PackedArrays 

The latter points are implementations of my proposal, https://github.com/godotengine/godot-proposals/issues/11722, and can be pared back if desired.

ColinSORourke committed:
Added Find Sequence to Spans

Added FindSequence to Span.h

Refactored String find to use Span::Find_Seq in Ustring.cpp

Ivorforce commented on core/templates/span.h:
```diff
@@ -84,6 +86,37 @@ constexpr int64_t Span<T>::find(const T &p_val, int64_t p_from) const {
 
+template <typename T>
+constexpr int64_t Span<T>::find_seq(const Span<T> &p_span, int64_t p_from) const {
+	const int64_t s = size();
+	const int64_t p_s = p_span.size();
+
+	if (p_from < 0 || s == 0) {
+		return -1;
+	}
+	if (p_s == 0 || p_s > s){
+		return -1;
+	}
+	if (p_s == 1){
+		return find(p_span.ptr()[0], p_from);
+	}
+
```
As explained, skipping this part:

```suggestion
```

Ivorforce commented on core/templates/span.h:
```diff
@@ -103,6 +136,45 @@ constexpr int64_t Span<T>::rfind(const T &p_val, int64_t p_from) const {
 
+template <typename T>
+constexpr int64_t Span<T>::rfind_seq(const Span<T> &p_span, int64_t p_from) const {
+	const int64_t s = size();
+	const int64_t p_s = p_span.size();
```
I think both variables would be better inlined (aka use `size()` instead of `s` everywhere). The `size()` function is guaranteed to be inlined, so it's not a performance concern, and I think `size()` is more readable than `s` :)

Ivorforce commented on doc/classes/PackedInt32Array.xml:
```diff
@@ -160,6 +168,14 @@
+			<param index="0" name="sequence" type="PackedByteArray" />
+			<param index="1" name="from" type="int" default="0" />
+			<description>
+				Searches the array in reverse-order for a sub-sequence.  Optionally, a start search index can be passed. If negative, the start index is considered relative to the end of the array.
```
> Searches the array in reverse-order for a sub-sequence.

Generally I think this is a good formulation, but `String` refers to 'subsequence' with a different definition (in the function `is_subsequence_of` which is also exposed):
```gdscript
assert("abc".is_subsequence_of("a1b1c1"))
```
I had to look up the term, and I think `String`'s definition for it is correct. Perhaps let's just formulate it like:
```suggestion
				Searches the array in reverse-order for the occurrence of [param sequence].  Optionally, a start search index can be passed. If negative, the start index is considered relative to the end of the array.
```

ColinSORourke replied:
Adjusted wording here, I didn't know "Subsequence" was its own distinct term. Finalized on "Searches ... for an occurrence of the given sequence." which I think reads well.
<img width="881" alt="Screenshot 2025-03-18 at 7 19 46 PM" src="https://github.com/user-attachments/assets/2438dc29-ba89-41f5-8524-a4a959e488a7" />

Ivorforce reviewed (commented):
Hi, this is looking very good!
I'm personally in favour of this change. When dealing with `Packed*Arrays` I found them to be a bit lacking in terms of exposed functionality. This is a welcome addition in my mind!

The implementation looks pretty optimal to me, and I think it feels right at home in `Span`.

However, I'm trying to establish `Span` to be unopinionated in another PR (https://github.com/godotengine/godot/pull/104286). 
If that gets merged, it would mean that negative or out-of-bounds index checks should be performed outside of `Span`, in the user facing functions (causing some duplicate code in `Vector` and `String` separately). My reasoning for this is because opinionated code should be explicit.
The other PR is accepted for a merge, so unless someone has concerns, we shouldn't have to wait long for that :)

The name `find_seq` and documentation work for me.

ColinSORourke commented:
Committed just now with adjustments based on Ivorforce's feedback! Tested it all still works too. Let me know if anyone else has thoughts, or if this is looking good!

Ivorforce reviewed (commented):
Looking at this again, I'm very happy with the decision to have the containers do the logic and `Span` do the dumb but fast implementation. This feels sane to me.

I'd like to ask you to harmonize the two functions though, according to adjustments made in https://github.com/godotengine/godot/pull/104286/ regarding boundary conditions. In particular:
- Negative indices from `-1` to `-size() -1` are indices starting from the back
- Negative indices beyond that and indices above `size()` are incorrect indices and return -1

This will change behavior in the `String` implementation, but I think it's worth it for consistency in the long run.

ColinSORourke commented:
Newest commit synchronizes all the boundary behaviors, and improves comments, like suggested

Ivorforce reviewed (approved):
Awesome, it's looking perfect now! Thanks for indulging all my feedback :)

Just needs a rebase onto `master` because my other PR got merged.
Also, you'll need to re-run `doctool` because the `default` values in the `xml` are out of sync (still 0 for `rfind_seq` instead of `-1`).

Finally, I think it would be good to hear one or two more voices on whether we should add / expose this feature in the first place. I personally think so, but I'm not the only or most relevant voice for this :)

Ivorforce commented on core/templates/span.h:
```diff
@@ -64,9 +64,12 @@ class Span {
 
 	// Algorithms.
-	constexpr int64_t find(const T &p_val, uint64_t p_from = 0) const;
+	constexpr int64_t find(const T &p_val, int64_t p_from = 0) const;
```
I think you accidentally swapped types here

Ivorforce reviewed (changes requested):
Just bringing a quick update: We've found a problem with negative indexes, and were trying to figure out together on chat how best to handle this.

I'll summarize it quickly here:
```
# What should this do?
"ABCDEF".rfind_seq("CD", -2)
```

We see 3 options:
- Old behavior: recognize as "invalid index" and default to normal search, i.e. iterate from index `4 to 0`
- Interpret `-2` as an index into the array, but clamp to a valid search index, i.e. iterate from `4 to 0`
- Interpret `-2` as an index into the search space, i.e. iterate from `3 to 0`

Personally, I think the correct choice would be to use the third option. It's consistent with forwards and backwards iterating logic, basically using back anchors for back indexes and front anchors for forward indexes. The search space is the following:

```
CD
ABCDEF

 CD
ABCDEF

  CD
ABCDEF

   CD
ABCDEF

    CD
ABCDEF
```

A value of `-1` can be interpreted as "start from the back", while `-2` can be interpreted as "start from the back, but skip one index", etc.

I can see that it's a bit confusing, but it generalizes to size 1 behavior, and is consistent with theoretical equal slicing behavior:
```
# -0 indexing does not exist, but hopefully you get the gist:
"ABCDEF"[:].rfind_seq("ABC") == "ABCDEF".rfind_seq("ABC", -1)
"ABCDEF"[:-1].rfind_seq("ABC") == "ABCDEF".rfind_seq("ABC", -2)
"ABCDEF"[:-2].rfind_seq("ABC") == "ABCDEF".rfind_seq("ABC", -3)
# [...]
```

It would be interesting to hear more opinions on this. It's certainly not a trivial problem!
But I think we should probably not merge this until we can settle on a behavior.

ColinSORourke commented:
I've been chatting with Ivor about this too, it's a tough cookie.

I'm personally in favor of interpreting negative values of P-From as direct indexes into the array rather than the search space.
1. I think it's unintuitive that String.rfind(X, -5) would start iteration at at a different index depending on the length of X
2. I think clamping valid array indexes down to a limit makes sense because it's not actually interpreting user input, simply skipping iterations we know cannot be correct & avoiding out of bound checks.
3. I think direct index is more consistent with the old behavior and more consistent with other find functions, including with other similar languages like Python.

Table here includes examples of the iteration start points for the 3 different options across several cases:
<img width="1135" alt="Screenshot 2025-03-19 at 6 44 44 PM" src="https://github.com/user-attachments/assets/d3911316-5693-40ad-80a5-11399e7216d9" />

ColinSORourke commented:
> I noticed this is exposed to _all_ PackedArray types in scripting, but it always takes a PackedByteArray as parameter (even in e.g. PackedColorArray). Shouldn't it take a parameter of the same type and convert it behind the scenes?

I think that's an error in the Documentation and not actual behavior, a result of me copy pasting in the Docs before I knew DocTools existed! Let me get that fixed!

aaronfranke commented on core/templates/vector.h:
```diff
@@ -114,6 +115,26 @@ class Vector {
 		return span().find(p_val, p_from);
 	}
+
+	Size find_seq(const Vector<T> &p_vector, Size p_from = 0) const {
```
The name `find_sequence` would be better than `find_seq`. This is not a super common function so it should not be shortened.

Ivorforce commented on core/string/ustring.cpp:
```diff
@@ -3243,46 +3243,22 @@ String String::substr(int p_from, int p_chars) const {
-
-	if (src_len == 0 || len == 0) {
-		return -1; // won't find anything!
+	if (p_from < 0 || p_from >= length()) {
```
Needs to account for large starting indices:

```suggestion
	if (p_from < 0 || p_from + p_str.length() > length() || p_str.is_empty()) {
```

With this, the next `if` block should be deletable entirely, if I'm not mistaken.
Same for `rfind` implementation.

Ivorforce commented on scene/gui/text_edit.cpp:
```diff
@@ -4403,7 +4403,7 @@ Point2i TextEdit::search(const String &p_key, uint32_t p_search_flags, int p_fro
 		pos = -1;
 
-		int pos_from = (p_search_flags & SEARCH_BACKWARDS) ? text_line.length() : 0;
+		int pos_from = (p_search_flags & SEARCH_BACKWARDS) ? text_line.length() - 1 : 0;
```
```suggestion
		int pos_from = (p_search_flags & SEARCH_BACKWARDS) ? - 1 : 0;
```

Ivorforce commented on core/string/ustring.cpp:
```diff
@@ -3243,46 +3243,22 @@ String String::substr(int p_from, int p_chars) const {
 int String::find(const String &p_str, int p_from) const {
 	if (p_from < 0) {
-		return -1;
+		p_from = length() + p_from;
```
```suggestion
		p_from = length() - p_str.length() + (int64_t)p_from;
```

Ivorforce reviewed (commented):
Hi! Sorry for getting back to you so late. This is a hard decision to make indeed!

I finally got around to thinking about this again, but I reached the same conclusion as I did earlier - the only 'consistent' behavior would be to make `-1` the "last possible index in the search space". Making it the last index in the 'search string' would require more complex logic, for questionable benefit. I really tried to like your suggestion, but it doesn't work for me. I hope that's ok!

Let try to get this merged (or at least mergable, someone may still object to this being added) in time for 4.5 :)

I think there's still a hole in the logic; a few small changes should fix this.

Ivorforce commented on core/string/ustring.cpp:
```diff
@@ -3083,59 +3083,28 @@ String String::substr(int p_from, int p_chars) const {
-
 	const int src_len = strlen(p_str);
 
-	const int len = length();
```
I'm not 100% sure whether `length()` can be optimized out in `String`, since it's in RAM rather than on the stack. It may be better to just keep it in `const int len`, should also reduce your line change count again :)

Ivorforce reviewed (approved):
Just one more thing, should probably keep caching `length()` values for inner loops.

Otherwise, it's perfect now, in my opinion! Thanks for all the hard work.

We'll probably need at least one more opinion before we merge this.
In particular, it's interesting to know:

- Whether we want to expose this function (see https://github.com/godotengine/godot-proposals/issues/11722).
- Whether we're ready to change all `find` functions' negative indexing (from "fail and / or lax" to consistent negative indexing behavior).
- Whether the wording in the docs is appropriate.

I'd like to point out that we discussed this change in the core meeting, and while it seemed the feature may be useful, there were no other opinions on implementation details.

ColinSORourke commented:
For the purposes of documentation, here are some examples of the behaviors I think the docs should clarify.

```
“ABCDEF”.find(“DEF”, -1)
// Old Behavior: Don't search, return -1
// New Behavior: Search from D->D, return 3

"ABCDEF".find("BCD", -2)
// Old behavior: Don't search, return -1
// New Behavior: Search C->D, return -1

"ABCDEF".find("BCDE", -2)
// Old behavior: Don't search, return -1
// New Behavior: Search B->C, return 1

"ABCDEF".find("BCD", -3)
// Old behavior: Don't search, return -1
// New Behavior: Search B->D, return 1

"ABCDEF".find("BCD", -5)
// Old Behavior: Don't search, return -1
// New Behavior: Don't search, return -1

“ABCDEF”.rfind(“DEF”, -2)
// Old behavior: Search D->A, return 3
// New Behavior: Search C->A, return -1

"ABCDEF".rfind("CDEF", -2)
// Old behavior: Search C->A, return 2
// New Behavior: Search B->A, return -1

"ABCDEF".rfind("DEF", 4)
// Old Behavior: Search D->A, return 3
// New Behavior: Don't Search, return -1

"ABCDEF".rfind("DEF", 6)
// Old Behavior: Search D->A, return 3
// New Behavior: Don't search, return -1
```

Ivorforce commented on core/string/ustring.cpp:
```diff
@@ -3325,85 +3298,44 @@ int String::findn(const char *p_str, int p_from) const {
-	if (limit < 0) {
-		return -1;
-	}
+	const int src_len = strlen(p_str);
+	const int len = length();
```
I don't think `src_len` is a good name for the length of the given string. You can stay agnostic, but if you want to rename them, I'd rename them all uniformly to this scheme:

```suggestion
	const int str_len = strlen(p_str);
	const int len = length();
```

ColinSORourke replied:
Changed, thanks!

fire commented:
> For the purposes of documentation, here are some examples of the behaviors I think the docs should clarify. These all relate to both Find & Rfind interpreting negative values of from in conjunction with what's length, and rfind now being stricter by returning -1 for certain out-of-bounds values of from.

If possible I would try to encode these as unit tests. Documentation is a bit weaker that code tests.

fire commented:
Also I would suggest if we redefine find and rfind, they should have new names.

Ivorforce commented:
> Also I would suggest if we redefine find and rfind, they should have new names.

The behavior will be the same for most people. Only negative (beyond -1) and invalid index behavior changes. It's unlikely many people were relying on the old behavior, i think.

Ivorforce commented:
We discussed this recently in a core meeting. It was decided not to expose the `find_sequence` functions at this time, because the demand is unclear. It is preferable to make a GDextension with `find_sequence` as a feature.
The refactors are welcome, however, so the PR just needs to be downsized a bit.

ColinSORourke commented:
(6/17) - I've pared back this PR to just adding Find_Sequence to Span.H, and refactoring UString.cpp to make use of this on the backend.

I'll be working on a GDExtension to implement my proposal, and will share here when that is complete

Ivorforce reviewed (approved):
Looks great to me! Thank you again for going through all these iterations.
We're past the 4.5 feature freeze now, so let's aim for 4.6 dev 1 to merge this. It would be nice to get final docs feedback on the `.xml` changes too.

ColinSORourke commented:
Busy few months for me, but I had time tonight and wrote a Find_Sequence GDExtension in C++. It performs much worse than a native Find_Sequence implementation, and I would like to return to my proposal of adding this function natively to the engine.

Methodology: I rebased this PR onto master, syncing it to the most recent builds of 4.5. I added a Find_Sequence to Core/Templates/Vector.h, and exposed it to PackedByteArray in Core/Variant/Variant_call.cpp. The GDExtension was written in C++, and implements the same search algorithm, taking two PackedByteArrays as input.

100 trials searching 134.2 MB file (retrieved via FileAccess.get_file_as_bytes()) for a Byte Sequence occurring early in the file. I tested
1. A GDScript implementation of the algorithm
2. Using the Native String.find() on a Hexadecimal String of the file (which calls the backend Span.find_sequence)
3. Using a Native PackedByteArray.find_sequence (implemented to directly call backend Span.find_sequence)
4. Using the C++ GDExtension find_sequence

Results:
1. GDScript Find Sequence elapsed 8.1967830657959 
2. String Find Sequence elapsed 1.37343001365662 
3. Native Find Sequence elapsed 0.26774501800537
4. GDExtension Find Sequence elapsed 4.45171809196472

I've attached some screenshots of the relevant code, I believe I wrote each behavior as optimally as I could.

<img width="915" height="734" alt="FindSeqPerfTest" src="https://github.com/user-attachments/assets/85192a8c-3277-4509-ac72-3af8bf3e3262" /> 
<img width="570" height="296" alt="GDScriptFindSeq" src="https://github.com/user-attachments/assets/4c27042c-33b5-45a8-a859-225d8f70fcba" />
<img width="800" height="618" alt="GDExtensionFindSeq" src="https://github.com/user-attachments/assets/a1234df4-b3db-4039-af79-458b39bac9d5" />

ColinSORourke commented:
Can confirm @Ivorforce was correct here - appropriate Extension code resolved the performance difference, no change has been made to this PR :)

Repiteo commented:
Thanks!

AgostinoSturaro commented:
@ColinSORourke  Hi, I found this thread in the 4.6 dev 1 release notes and got curious about it.
However, I cannot figure out the examples.

> For the purposes of documentation, here are some examples of the behaviors I think the docs should clarify. These all relate to both Find & Rfind interpreting negative values of from in conjunction with what's length, and rfind now being stricter by returning -1 for certain out-of-bounds values of from.
> 
> ```
> “ABCDEF”.find(“DEF”, -1)
> // Old Behavior: Don't search, return -1
> // New Behavior: Search from D->D, return 3
> 
> "ABCDEF".find("BCD", -2)
> // Old behavior: Don't search, return -1
> // New Behavior: Search C->D, return -1
> 
> "ABCDEF".find("BCDE", -2)
> // Old behavior: Don't search, return -1
> // New Behavior: Search B->C, return 1
> 
> "ABCDEF".find("BCD", -3)
> // Old behavior: Don't search, return -1
> // New Behavior: Search B->D, return 1
> 
> "ABCDEF".find("BCD", -5) // A user might think -5 is in bounds as an index, but it is not.
> // Old Behavior: Don't search, return -1
> // New Behavior: Don't search, return -1
> ...
> ```

If I do the same in Python, I completely different results, mostly `-1`.
In that language, a negative index (as I understand it) means "start from the end" of the string/list you are searching into. E.g.
```
print(f'{"ABCD".find("CD", -1)}')
-1
print(f'{"ABCD".find("CD", -2)}')
2
```
Isn't it the same in GDScript? My understanding was the languages are supposed to be similar.

Thanks for your patience.

Ivorforce commented:
> @Ivorforce So, by "the last possible position to search" you actually mean "the last possible position a **successful** search can start from".

Yes, this is a good way to formulate that!

> I didn't understand the length of both strings should be considered. Python only considers the first string, even for negative indexes.

That's the same perspective that @ColinSORourke has about this as well. I think about it differently, but as long as the behavior is clear, I think it doesn't matter so much.

> The example above, plus your explanation, clarified it for me. Can it be added to the doc? Thank you!

Would you be open to create that PR? As someone who freshly had to understand it, I think your perspective would be very valuable for finding the right way to put it in the docs. I'm happy to place a review on it when it's there!
